### PR TITLE
Make extension unloading atomic

### LINKED ISSUE #6113: Exception in cog_unload stops module cleanup
## Summary

Throwing an error in a cog's `cog_unload` method leaves the cog's module to stay in `sys.modules` and the internal `Bot.__extensions` list. This means that the cog is stuck in a state where it can't be loaded (since the module it resides in is "still loaded") until the `unload_extension` is called for a second time to clean up. (For `n` cogs in an extension raising errors, the extension must be unloaded `n + 1` times for the module to be fully unloaded.)

## Reproduction Steps

in `cogs/BadCog.py`:

```py
from discord.ext import commands

class BadCog(commands.Cog):
    def cog_unload(self):
        raise Exception()

    @commands.command()
    async def foo(self, ctx):
        await ctx.send("still loaded!")

def setup(bot):
    bot.add_cog(BadCog())
```

Using [jishaku](https://github.com/Gorialis/jishaku/tree/master/jishaku):

?jsk load cogs.BadCog
:inbox_tray: `cogs.BadCog`

?foo
still loaded!

?jsk unload cogs.BadCog
:outbox_tray::warning: cogs.BadCog`

?foo << this message receives no response

?jsk unload cogs.BadCog
:outbox_tray: cogs.BadCog

## Expected Results

One of the following:
- Proper cleanup ignoring exceptions in either `Cog._eject`, `Bot.remove_cog`, `Bot._remove_module_references`, or `Bot.unload_extension`
- Atomic unloading, meaning that Cogs are re-loaded to the bot if unloading the extension fails.

## Actual Results

I am sad.

## Intents

N/A.

## Checklist

- [x] I have searched the open issues for duplicates.
- [x] I have shown the entire traceback, if possible.
- [x] I have removed my token from display, if visible.
- [x] I have provided the intents that my bot is using.

## System Information

N/A.

### PR DESCRIPTION
**Pull request by SebbyLaw**
## Summary

Closes #6113 

edit: talked with Danny on discord briefly. Extension unloading isn't meant to be atomic, but hasn't looked at #6113 yet.

## Checklist

- [x] If code changes were made then they have been tested.
    - [N/A] I have updated the documentation to reflect the changes.
- [x] This PR fixes an issue.
- [N/A] This PR adds something new (e.g. new method or parameters).
- [N/A] This PR is a breaking change (e.g. methods or parameters removed/renamed)
- [N/A] This PR is **not** a code change (e.g. documentation, README, ...)

### CODE REVIEW

File: discord/ext/commands/bot.py [hunk 1]
```diff
@@ -559,9 +559,19 @@ def cogs(self):
...
         for cogname, cog in self.__cogs.copy().items():
             if _is_submodule(name, cog.__module__):
-                self.remove_cog(cogname)
+                try:
+                    self.remove_cog(cogname)
+                except Exception as exc:
+                    # a cog raised an exception in cog_unload
+                    # add all cogs back to the module and propagate the exception
+                    for removed_cog in removed_cogs:
+                        self.add_cog(remove_cog)
```
Comment (Silari) [inline discord/ext/commands/bot.py:571]: Should be 'self.add_cog(removed_cog)'

File: discord/ext/commands/bot.py [hunk 2]
```diff
@@ -559,9 +559,18 @@ def cogs(self):
...
         # find all references to the module
         # remove the cogs registered from the module
+        removed_cogs = []
         for cogname, cog in self.__cogs.copy().items():
             if _is_submodule(name, cog.__module__):
-                self.remove_cog(cogname)
+                try:
+                    removed_cogs.append(cog)
+                    self.remove_cog(cogname)
+                except Exception as exc:
```
Comment (ShineyDev) [inline discord/ext/commands/bot.py:568]:
```suggestion
                except BaseException as exc:
```

**Review by ShineyDev — CHANGES_REQUESTED** [review]

File: discord/ext/commands/bot.py [hunk 3]
```diff
@@ -559,9 +559,18 @@ def cogs(self):
...
         # find all references to the module
         # remove the cogs registered from the module
+        removed_cogs = []
         for cogname, cog in self.__cogs.copy().items():
             if _is_submodule(name, cog.__module__):
-                self.remove_cog(cogname)
+                try:
+                    removed_cogs.append(cog)
+                    self.remove_cog(cogname)
+                except BaseException as exc:
```
Comment (mikeshardmind) [inline discord/ext/commands/bot.py:568]: Catching BaseException will interfere with things this should not interfere with should they happen at the wrong time.

**Review by mikeshardmind — CHANGES_REQUESTED** [review]

File: discord/ext/commands/bot.py (hunk 3 above)
Comment (SebbyLaw) [reply discord/ext/commands/bot.py]: After looking into it a bit more in depth, I agree.

**Comment by Rapptz** [comment]
I think there's a better approach here

**Pull request closed without merging**